### PR TITLE
XDP Testcases: Updated Compatible kernel version check

### PR DESCRIPTION
**Pull request by abhimarathe**
1. Changed check from kernel version based to feature (XDP_DROP counter) check.
2. Removed TestParameters from test xml and defined them in test script.

```
[LISAv2 Test Results Summary]
Test Run On           : 07/27/2020 19:25:29
ARM Image Under Test  : Canonical : UbuntuServer : 19.10-DAILY : latest
Initial Kernel Version: 5.4.0-1017-azure
Final Kernel Version  : 5.4.0-1017-azure
Total Test Cases      : 1 (1 Passed, 0 Failed, 0 Aborted, 0 Skipped)
Total Time (dd:hh:mm) : 0:0:12

   ID TestArea             TestCaseName                                                                TestResult TestDuration(in minutes) 
-------------------------------------------------------------------------------------------------------------------------------------------
    1 XDP                  VERIFY-SRIOV-FAILSAFE-XDP                                                         PASS                11.22 
      ARMImageName: Canonical UbuntuServer 19.10-DAILY latest, Networking: SRIOV, OverrideVMSize: Standard_DS4_v2, SetupType: TwoVM2NIC, TestLocation: eastus2


Logs can be found at C:\LISAv2\YK43\TestResults\2020-07-27-12-24-37-2914

[LISAv2 Test Results Summary]
Test Run On           : 07/30/2020 18:01:53
ARM Image Under Test  : Canonical : UbuntuServer : 19.10-DAILY : latest
Initial Kernel Version: 5.4.0-1017-azure
Final Kernel Version  : 5.4.0-1017-azure
Total Test Cases      : 1 (1 Passed, 0 Failed, 0 Aborted, 0 Skipped)
Total Time (dd:hh:mm) : 0:0:9

   ID TestArea             TestCaseName                                                                TestResult TestDuration(in minutes)
-------------------------------------------------------------------------------------------------------------------------------------------
    1 XDP                  VERIFY-XDP-LOAD-UNLOAD                                                            PASS                  7.2
      ARMImageName: Canonical UbuntuServer 19.10-DAILY latest, Networking: SRIOV, OverrideVMSize: Standard_DS4_v2, SetupType: TwoVM2NIC, TestLocation: eastus2


Logs can be found at C:\LISAv2\WN38\TestResults\2020-07-30-11-00-09-3378

[LISAv2 Test Results Summary]
Test Run On           : 07/30/2020 18:14:54
ARM Image Under Test  : Canonical : UbuntuServer : 19.10-DAILY : latest
Initial Kernel Version: 5.4.0-1017-azure
Final Kernel Version  : 5.4.0-1017-azure
Total Test Cases      : 1 (1 Passed, 0 Failed, 0 Aborted, 0 Skipped)
Total Time (dd:hh:mm) : 0:0:6

   ID TestArea             TestCaseName                                                                TestResult TestDuration(in minutes)
-------------------------------------------------------------------------------------------------------------------------------------------
    1 XDP                  VERIFY-XDP-ACTION-DROP                                                            PASS                 4.58
      ARMImageName: Canonical UbuntuServer 19.10-DAILY latest, Networking: SRIOV, OverrideVMSize: Standard_DS4_v2, SetupType: TwoVM2NIC, TestLocation: eastus2


Logs can be found at C:\LISAv2\SR28\TestResults\2020-07-30-11-13-55-2624
[LISAv2 Test Results Summary]
Test Run On           : 07/30/2020 18:23:12
ARM Image Under Test  : Canonical : UbuntuServer : 19.10-DAILY : latest
Initial Kernel Version: 5.4.0-1017-azure
Final Kernel Version  : 5.4.0-1017-azure
Total Test Cases      : 3 (3 Passed, 0 Failed, 0 Aborted, 0 Skipped)
Total Time (dd:hh:mm) : 0:0:12

   ID TestArea             TestCaseName                                                                TestResult TestDuration(in minutes)
-------------------------------------------------------------------------------------------------------------------------------------------
    1 XDP                  VERIFY-XDP-ACTION-TX                                                              PASS                 4.93
      ARMImageName: Canonical UbuntuServer 19.10-DAILY latest, Networking: SRIOV, OverrideVMSize: Standard_DS4_v2, SetupType: TwoVM2NIC, TestLocation: eastus2
    2 XDP                  VERIFY-XDP-ACTION-ABORTED                                                         PASS                 4.33
      ARMImageName: Canonical UbuntuServer 19.10-DAILY latest, Networking: SRIOV, OverrideVMSize: Standard_DS4_v2, SetupType: TwoVM2NIC, TestLocation: eastus2
    3 XDP                  VERIFY-XDP-MTU                                                                    PASS                 4.78
      ARMImageName: Canonical UbuntuServer 19.10-DAILY latest, Networking: SRIOV, OverrideVMSize: Standard_DS4_v2, SetupType: TwoVM2NIC, TestLocation: eastus2


Logs can be found at C:\LISAv2\DJ37\TestResults\2020-07-30-11-22-10-5955
```

### CODE REVIEW

**Comment by abhimarathe** [comment]
```
[LISAv2 Test Results Summary]
Test Run On           : 07/30/2020 20:13:03
ARM Image Under Test  : Canonical : UbuntuServer : 19.10-DAILY : latest
Initial Kernel Version: 5.4.0-1017-azure
Final Kernel Version  : 5.4.0-1017-azure
Total Test Cases      : 1 (1 Passed, 0 Failed, 0 Aborted, 0 Skipped)
Total Time (dd:hh:mm) : 0:0:10

   ID TestArea             TestCaseName                                                                TestResult TestDuration(in minutes)
-------------------------------------------------------------------------------------------------------------------------------------------
    1 XDP                  VERIFY-SRIOV-FAILSAFE-XDP                                                         PASS                 9.06
      ARMImageName: Canonical UbuntuServer 19.10-DAILY latest, Networking: SRIOV, OverrideVMSize: Standard_DS4_v2, SetupType: TwoVM2NIC, TestLocation: eastus2


Logs can be found at C:\LISAv2\QP53\TestResults\2020-07-30-13-12-10-6711
```

**Review by juhlee-microsoft — DISMISSED** [review]

File: Testscripts/Windows/VERIFY-XDP-ACTION.ps1 [hunk 1]
```diff
@@ -47,21 +45,13 @@ function Main {
...
-                return $global:ResultSkipped
-            }
+        $numberOfXDPQueues = Run-LinuxCmd -ip $receiverVMData.PublicIP -port $receiverVMData.SSHPort `
+                -username $user -password $password -command "ethtool -S $iFaceName  | grep xdp_drop | wc -l"
+        Write-LogInfo "Number of XDP DROP queues: $numberOfXDPQueues"
+        if ($numberOfXDPQueues -gt 0) {
+            Write-LogInfo "Kernel version supports XDP"
         } else {
-            Write-LogInfo "Unsupported distro: $($global:DetectedDistro)."
+            Write-LogWarn "Unsupported Kernel or unsupported distro $($global:DetectedDistro)"
```
Comment (harz566) [inline Testscripts/Windows/VERIFY-XDP-ACTION.ps1]: can we add the actual kernel info in the Warning message?  $global:initialkernelversion

Comment (abhimarathe) [reply Testscripts/Windows/VERIFY-XDP-ACTION.ps1]: fixed.

**Review by harz566 — DISMISSED** [review]

**Review by juhlee-microsoft — DISMISSED** [review]

File: Testscripts/Windows/VERIFY-XDP-ACTION.ps1 [hunk 2]
```diff
@@ -10,8 +10,6 @@
 param([object] $AllVmData,
     [object] $CurrentTestData)
 
-$MIN_KERNEL_VERSION = "5.6"
-$RHEL_MIN_KERNEL_VERSION = "4.18.0-213"
 $iFaceName = "eth1"
```
Comment (johnsongeorge-w) [inline Testscripts/Windows/VERIFY-XDP-ACTION.ps1:13]: Is it always eth1

File: Testscripts/Windows/VERIFY-XDP-MULTIPLE-NICS.ps1 [hunk 3]
```diff
@@ -69,21 +67,13 @@ function Main {
...
-                return $global:ResultSkipped
-            }
+        $numberOfXDPQueues = Run-LinuxCmd -ip $clientVMData.PublicIP -port $clientVMData.SSHPort `
+                -username $user -password $password -command "ethtool -S $iFaceName  | grep xdp_drop | wc -l"
+        Write-LogInfo "Number of XDP DROP queues: $numberOfXDPQueues"
+        if ($numberOfXDPQueues -gt 0) {
+            Write-LogInfo "Kernel version supports XDP"
         } else {
-            Write-LogInfo "Unsupported distro: $($global:DetectedDistro)."
+            Write-LogWarn "Unsupported Kernel $($global:initialkernelversion) or unsupported distro $($global:DetectedDistro)"
```
Comment (johnsongeorge-w) [inline Testscripts/Windows/VERIFY-XDP-MULTIPLE-NICS.ps1]: Can we just print ```XDP not supported``` ? Same for all occurrences

**Review by johnsongeorge-w — CHANGES_REQUESTED** [review]

File: Testscripts/Windows/VERIFY-XDP-ACTION.ps1 (hunk 2 above)
Comment (abhimarathe) [reply Testscripts/Windows/VERIFY-XDP-ACTION.ps1:13]: With the testcase config we have added an entra nic. With my observation and which is backed by @haiyangz about the name of Synthetic nic would always follow eth`n` format. VF name that is not stable but we can rely on Synthetic NIC

File: Testscripts/Windows/VERIFY-XDP-MULTIPLE-NICS.ps1 (hunk 3 above)
Comment (abhimarathe) [reply Testscripts/Windows/VERIFY-XDP-MULTIPLE-NICS.ps1]: Giving kernel version is a valid point as this is a new feature and most of the distro version does not have it, mostly we have to run this with -CustomKernel parameter and to check which kernel version we have ran compatibility check on, it gives output right away.

**Comment by abhimarathe** [comment]
```
[LISAv2 Test Results Summary]
Test Run On           : 07/31/2020 22:08:02
ARM Image Under Test  : Canonical : UbuntuServer : 19.10-DAILY : latest
Initial Kernel Version: 5.4.0-1017-azure
Final Kernel Version  : 5.4.0-1017-azure
Total Test Cases      : 3 (3 Passed, 0 Failed, 0 Aborted, 0 Skipped)
Total Time (dd:hh:mm) : 0:0:15

   ID TestArea             TestCaseName                                                                TestResult TestDuration(in minutes)
-------------------------------------------------------------------------------------------------------------------------------------------
    1 XDP                  VERIFY-XDP-ACTION-DROP                                                            PASS                 4.46
      ARMImageName: Canonical UbuntuServer 19.10-DAILY latest, Networking: SRIOV, OverrideVMSize: Standard_DS4_v2, SetupType: TwoVM2NIC, TestLocation: eastus2
    2 XDP                  VERIFY-XDP-ACTION-TX                                                              PASS                 4.69
      ARMImageName: Canonical UbuntuServer 19.10-DAILY latest, Networking: SRIOV, OverrideVMSize: Standard_DS4_v2, SetupType: TwoVM2NIC, TestLocation: eastus2
    3 XDP                  VERIFY-XDP-ACTION-ABORTED                                                         PASS                 4.36
      ARMImageName: Canonical UbuntuServer 19.10-DAILY latest, Networking: SRIOV, OverrideVMSize: Standard_DS4_v2, SetupType: TwoVM2NIC, TestLocation: eastus2


Logs can be found at C:\LISAv2\RY53\TestResults\2020-07-31-15-06-59-9447
[LISAv2 Test Results Summary]
Test Run On           : 07/31/2020 21:42:01
ARM Image Under Test  : Canonical : UbuntuServer : 19.10-DAILY : latest
Initial Kernel Version: 5.4.0-1017-azure
Final Kernel Version  : 5.4.0-1017-azure
Total Test Cases      : 1 (1 Passed, 0 Failed, 0 Aborted, 0 Skipped)
Total Time (dd:hh:mm) : 0:0:4

   ID TestArea             TestCaseName                                                                TestResult TestDuration(in minutes)
-------------------------------------------------------------------------------------------------------------------------------------------
    1 XDP                  VERIFY-XDP-COMPLIANCE                                                             PASS                 3.81
      ARMImageName: Canonical UbuntuServer 19.10-DAILY latest, Networking: SRIOV, OverrideVMSize: Standard_DS4_v2, SetupType: OneVM2NIC, TestLocation: eastus2


Logs can be found at C:\LISAv2\GJ62\TestResults\2020-07-31-14-41-04-0585


```

**Review by harz566 — DISMISSED** [review]

**Review by juhlee-microsoft — DISMISSED** [review]

File: Testscripts/Linux/XDPDumpSetup.sh [hunk 4]
```diff
@@ -110,6 +112,25 @@ function Run_XDPDump {
     fi
 }
 
+# Check if kernel supports XDP or not
+function kernel_check {
```
Comment (johnsongeorge-w) [inline Testscripts/Linux/XDPDumpSetup.sh]: Can we define it as check_xdp_support ?

File: Testscripts/Linux/XDPDumpSetup.sh [hunk 5]
```diff
@@ -110,6 +112,25 @@ function Run_XDPDump {
...
+    local install_ip="${1}"
+    local nic_name="${2}"
+    command="ethtool -S ${nic_name}  | grep xdp_drop | wc -l"
+    xdp_counter="$(ssh ${install_ip} $command)"
+    if [ $xdp_counter -gt 0 ]; then
+        LogMsg "Kernel version supports XDP"
+    else
+        LogErr "Kernel Version does not support XDP"
+        SetTestStateSkipped
+        exit 1
```
Comment (johnsongeorge-w) [inline Testscripts/Linux/XDPDumpSetup.sh]: Why exit with 1 ?

File: Testscripts/Windows/PERF-XDP-RX-DROP.ps1 [hunk 6]
```diff
@@ -92,7 +82,7 @@ collect_VM_properties
         # Terminate process if ran more than 5 mins
         # TODO: Check max installation time for other distros when added
         $timer = 0
-        while ((Get-Job -Id $testJob).State -eq "Running") {
+        while ($testJob -and ((Get-Job -Id $testJob).State -eq "Running")) {
```
Comment (johnsongeorge-w) [inline Testscripts/Windows/PERF-XDP-RX-DROP.ps1:85]: Why condition is enhanced. Are we fixing any other problem apart from clean up ?

File: Testscripts/Windows/VERIFY-SRIOV-FAILSAFE-XDP.ps1 [hunk 7]
```diff
@@ -43,7 +42,7 @@ function Ping-XDPDump {
     }
 
     $currentStatus = Run-LinuxCmd -ip $VMData.PublicIP -port $VMData.SSHPort -username $user -password $password `
-        -command "tail -1 ~/xdpdumpout$LogSuffix.txt | head -1" -runAsSudo
+        -command "tail -1 ~/xdpdumpout$LogSuffix.txt" -runAsSudo
```
Comment (johnsongeorge-w) [inline Testscripts/Windows/VERIFY-SRIOV-FAILSAFE-XDP.ps1:45]: ```head -1``` removed ?

File: Testscripts/Linux/XDPDumpSetup.sh (hunk 4 above)
Comment (johnsongeorge-w) [inline Testscripts/Linux/XDPDumpSetup.sh]: I think this function can be supported as an utility function. Can we define this in utils.sh or other common file so other test scripts can use it ?

**Review by johnsongeorge-w — CHANGES_REQUESTED** [review]

File: Testscripts/Linux/XDPDumpSetup.sh (hunk 4 above)
Comment (abhimarathe) [reply Testscripts/Linux/XDPDumpSetup.sh]: For the current xdp testcases, this script is executing to make sure XDP is running on the vm. I don't see any other benefit of adding this to utils.sh or any other common file.

File: Testscripts/Windows/VERIFY-SRIOV-FAILSAFE-XDP.ps1 (hunk 7 above)
Comment (abhimarathe) [reply Testscripts/Windows/VERIFY-SRIOV-FAILSAFE-XDP.ps1:45]: yes. As tail -1 would only return 1 line. we don't need head -1.

File: Testscripts/Windows/PERF-XDP-RX-DROP.ps1 (hunk 6 above)
Comment (abhimarathe) [reply Testscripts/Windows/PERF-XDP-RX-DROP.ps1:85]: No. We are just making sure that testJob is not empty.

File: Testscripts/Linux/XDPDumpSetup.sh (hunk 5 above)
Comment (abhimarathe) [reply Testscripts/Linux/XDPDumpSetup.sh]: to indicate general error. Is there any issue with it?

File: Testscripts/Linux/XDPDumpSetup.sh (hunk 4 above)
Comment (abhimarathe) [reply Testscripts/Linux/XDPDumpSetup.sh]: fixed

File: Testscripts/Linux/XDPDumpSetup.sh (hunk 5 above)
Comment (johnsongeorge-w) [reply Testscripts/Linux/XDPDumpSetup.sh]: As per design test case should always return 0 after setting appropriate state.

File: Testscripts/Linux/XDPDumpSetup.sh [hunk 8]
```diff
@@ -110,6 +112,25 @@ function Run_XDPDump {
     fi
 }
 
+# Check if kernel supports XDP or not
+function check_xdp_support {
```
Comment (johnsongeorge-w) [inline Testscripts/Linux/XDPDumpSetup.sh:116]: Can we define it as an utility function to be used by all other testscripts ?

File: Testscripts/Linux/XDPDumpSetup.sh [hunk 9]
```diff
@@ -110,6 +112,25 @@ function Run_XDPDump {
...
+    if [ -z "${1}" -o -z "${2}" ]; then
+        LogErr "ERROR: must provide install ip and NIC Name to Run_XDPDump"
+        SetTestStateAborted
+        exit 1
```
Comment (johnsongeorge-w) [inline Testscripts/Linux/XDPDumpSetup.sh]: If supporting as utility function then return. Don't exit and set state.

**Review by johnsongeorge-w — CHANGES_REQUESTED** [review]

File: Testscripts/Linux/XDPDumpSetup.sh (hunk 8 above)
Comment (abhimarathe) [reply Testscripts/Linux/XDPDumpSetup.sh:116]: As I mentioned above, This is a setup script and it is used by all xdp testscript. I don't see any necessary of adding it to utility

File: Testscripts/Linux/XDPDumpSetup.sh (hunk 9 above)
Comment (abhimarathe) [reply Testscripts/Linux/XDPDumpSetup.sh]: please refer to previous discussion.

**Comment by abhimarathe** [comment]
```
[LISAv2 Test Results Summary]
Test Run On           : 08/03/2020 22:34:51
ARM Image Under Test  : Canonical : UbuntuServer : 19.10-DAILY : latest
Initial Kernel Version: 5.3.0-1032-azure
Final Kernel Version  : 5.3.0-1032-azure
Total Test Cases      : 1 (0 Passed, 0 Failed, 0 Aborted, 1 Skipped)
Total Time (dd:hh:mm) : 0:0:3

   ID TestArea             TestCaseName                                                                TestResult TestDuration(in minutes)
-------------------------------------------------------------------------------------------------------------------------------------------
    1 XDP                  VERIFY-XDP-COMPLIANCE                                                          SKIPPED                 1.86
      ARMImageName: Canonical UbuntuServer 19.10-DAILY latest, Networking: SRIOV, OverrideVMSize: Standard_DS4_v2, SetupType: OneVM2NIC, TestLocation: eastus2


Logs can be found at C:\LISAv2\BQ50\TestResults\2020-08-03-15-33-54-9723
```

**Review by johnsongeorge-w — APPROVED** [review]

**Review by harz566 — APPROVED** [review]

File: Testscripts/Windows/VERIFY-SRIOV-FAILSAFE-XDP.ps1 [hunk 10]
```diff
@@ -85,25 +84,6 @@ function Main {
         Write-LogInfo "  SSH Port : $($serverVMData.SSHPort)"
         Write-LogInfo "  Internal IP : $($serverVMData.InternalIP)"
 
-        # Check for compatible kernel
-        $currentKernelVersion = Run-LinuxCmd -ip $clientVMData.PublicIP -port $clientVMData.SSHPort `
```
Comment (LiliDeng) [inline Testscripts/Windows/VERIFY-SRIOV-FAILSAFE-XDP.ps1:89]: where will the logic be handled? in check_xdp_support ?

**Review by LiliDeng — APPROVED** [review]

**Pull request merged**